### 1. Import the relevant libraries, along with your 'ords_prods_merge' dataframe.

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
# Assign target data to 'path' and verify

path = r'/Users/peterkim/Document/sql/CareerFoundry/Immersion Prep/4. Python Fundamentals for Data Analysts/Instacart Basket Analysis'
path

'/Users/peterkim/Document/sql/CareerFoundry/Immersion Prep/4. Python Fundamentals for Data Analysts/Instacart Basket Analysis'

In [3]:
# Importing dataframe; 'ordered_products_merged.pkl'

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ordered_products_merged.pkl'))


In [4]:
ords_prods_merge

order_id  user_id eval_set  order_number  order_dow  \
0          2539329        1    prior             1          2   
1          2398795        1    prior             2          3   
2           473747        1    prior             3          3   
3          2254736        1    prior             4          4   
4           431534        1    prior             5          4   
...            ...      ...      ...           ...        ...   
32434207   1320836   202557    prior            17          2   
32434208     31526   202557    prior            18          5   
32434209    758936   203436    prior             1          2   
32434210   2745165   203436    prior             2          3   
32434211   3093936   205420    prior             1          4   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         7                    15.0         196   
2                        12                    21.0         196   
3                         7                    29.0         196   
4                        15                    28.0         196   
...                     ...                     ...         ...   
32434207                 15                     1.0       43553   
32434208                 11                     3.0       43553   
32434209                  7                     NaN       42338   
32434210                  5                    15.0       42338   
32434211                 14                     NaN       28818   

          add_to_cart_order  reordered _merge  Unnamed: 0  \
0                         1          0   both         195   
1                         1          1   both         195   
2                         1          1   both         195   
3                         1          1   both         195   
4                         1          1   both         195   
...                     ...        ...    ...         ...   
32434207                  2          1   both       43557   
32434208                  2          1   both       43557   
32434209                  4          0   both       42342   
32434210                 16          1   both       42342   
32434211                  8          0   both       28820   

                           product_name  aisle_id  department_id  prices  
0                                  Soda        77              7     9.0  
1                                  Soda        77              7     9.0  
2                                  Soda        77              7     9.0  
3                                  Soda        77              7     9.0  
4                                  Soda        77              7     9.0  
...                                 ...       ...            ...     ...  
32434207            Orange Energy Shots        64              7     3.7  
32434208            Orange Energy Shots        64              7     3.7  
32434209          Zucchini Chips, Pesto        50             19     6.9  
32434210          Zucchini Chips, Pesto        50             19     6.9  
32434211  Hot Oatmeal Multigrain Raisin       130             14    10.3  

[32434212 rows x 16 columns]

### 2. Repeat the process to find the aggregated mean of the “order_number” grouped by “department_id” for a subset of the entire dataframe.

#### Creating a Subset & Running Data Checks

In [5]:
# Creating subset

df = ords_prods_merge[:1000000]

In [6]:
df.shape

(1000000, 16)

In [7]:
df

order_id  user_id eval_set  order_number  order_dow  \
0        2539329        1    prior             1          2   
1        2398795        1    prior             2          3   
2         473747        1    prior             3          3   
3        2254736        1    prior             4          4   
4         431534        1    prior             5          4   
...          ...      ...      ...           ...        ...   
999995   2836489   164627    prior            14          0   
999996   1843600   164632    prior             5          1   
999997    733106   164632    prior             9          2   
999998   1650124   164632    prior            17          6   
999999   3378258   164632    prior            22          6   

        order_hour_of_day  days_since_prior_order  product_id  \
0                       8                     NaN         196   
1                       7                    15.0         196   
2                      12                    21.0         196   
3                       7                    29.0         196   
4                      15                    28.0         196   
...                   ...                     ...         ...   
999995                 15                     6.0       30489   
999996                 19                     9.0       30489   
999997                 22                    10.0       30489   
999998                 17                    13.0       30489   
999999                 10                    15.0       30489   

        add_to_cart_order  reordered _merge  Unnamed: 0     product_name  \
0                       1          0   both         195             Soda   
1                       1          1   both         195             Soda   
2                       1          1   both         195             Soda   
3                       1          1   both         195             Soda   
4                       1          1   both         195             Soda   
...                   ...        ...    ...         ...              ...   
999995                  2          0   both       30491  Original Hummus   
999996                  2          0   both       30491  Original Hummus   
999997                  3          1   both       30491  Original Hummus   
999998                  1          1   both       30491  Original Hummus   
999999                  3          1   both       30491  Original Hummus   

        aisle_id  department_id  prices  
0             77              7     9.0  
1             77              7     9.0  
2             77              7     9.0  
3             77              7     9.0  
4             77              7     9.0  
...          ...            ...     ...  
999995        67             20     7.5  
999996        67             20     7.5  
999997        67             20     7.5  
999998        67             20     7.5  
999999        67             20     7.5  

[1000000 rows x 16 columns]

#### Grouping Data with Pandas

In [8]:
df.groupby('product_name')

#### Aggregating Data with egg()

- Performing a Single Aggregation

In [9]:
# Split the data into groups based on “department_id.”
# Apply the agg() function to each group to obtain the mean values for the “order_number” column.

df.groupby('department_id').agg({'order_number': ['mean']})


order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [10]:
# Alternative way of aggregation using 'mean()' instead of using 'agg()'

df.groupby('department_id')['order_number'].mean()


department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

In [11]:
# Instead of using square brackets, you could also set the column name off with a dot

df.groupby('department_id').order_number.mean()


department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

- Performing Multiple Aggregations

In [12]:
# Now, it includes two new arguments, min and max, after the mean

df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})


order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

### Aggregating Data with transform()

- Create a loyalty flag using loc

In [13]:
# Transform the ‘order_number’ column by applying the max() function from the NumPy library

ords_prods_merge['max_order'] = ords_prods_merge.groupby('user_id')['order_number'].transform(np.max)


In [14]:
ords_prods_merge.head(25)

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0    2539329        1    prior             1          2                  8   
1    2398795        1    prior             2          3                  7   
2     473747        1    prior             3          3                 12   
3    2254736        1    prior             4          4                  7   
4     431534        1    prior             5          4                 15   
5    3367565        1    prior             6          2                  7   
6     550135        1    prior             7          1                  9   
7    3108588        1    prior             8          1                 14   
8    2295261        1    prior             9          1                 16   
9    2550362        1    prior            10          4                  8   
10   2968173       15    prior            15          1                  9   
11   1870022       15    prior            17          2                 16   
12   1911383       15    prior            18          2                 11   
13   2715276       15    prior            21          1                  9   
14    487368       15    prior            22          1                 10   
15   2293453       19    prior             2          5                 14   
16   1973799       19    prior             5          6                 12   
17    532817       19    prior             7          4                 17   
18   1573906       21    prior            10          3                 10   
19   1593000       31    prior            10          3                  8   
20   2231262       31    prior            17          3                 11   
21   2580647       43    prior             6          4                 16   
22   2187180       43    prior             9          4                 12   
23   2497897       52    prior             1          1                  9   
24   1318871       52    prior             2          1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                      NaN         196                  1          0   both   
1                     15.0         196                  1          1   both   
2                     21.0         196                  1          1   both   
3                     29.0         196                  1          1   both   
4                     28.0         196                  1          1   both   
5                     19.0         196                  1          1   both   
6                     20.0         196                  1          1   both   
7                     14.0         196                  2          1   both   
8                      0.0         196                  4          1   both   
9                     30.0         196                  1          1   both   
10                     7.0         196                  2          0   both   
11                     8.0         196                  6          1   both   
12                     7.0         196                  1          1   both   
13                     7.0         196                  1          1   both   
14                    14.0         196                  1          1   both   
15                     6.0         196                  3          0   both   
16                     8.0         196                 15          1   both   
17                     6.0         196                  1          1   both   
18                     6.0         196                  2          0   both   
19                     7.0         196                 17          0   both   
20                     8.0         196                 14          1   both   
21                     4.0         196                  9          0   both   
22                     3.0         196                  1          1   both   
23                     NaN         196                  4          0   both   


### Deriving Columns with loc()

In [15]:
# Create a flag that assigns a “loyalty” label to a user ID based on its corresponding max order value

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)


loyalty_flag
Regular customer    15891077
Loyal customer      10293737
New customer         6249398
Name: count, dtype: int64

In [16]:
ords_prods_merge


order_id  user_id eval_set  order_number  order_dow  \
0          2539329        1    prior             1          2   
1          2398795        1    prior             2          3   
2           473747        1    prior             3          3   
3          2254736        1    prior             4          4   
4           431534        1    prior             5          4   
...            ...      ...      ...           ...        ...   
32434207   1320836   202557    prior            17          2   
32434208     31526   202557    prior            18          5   
32434209    758936   203436    prior             1          2   
32434210   2745165   203436    prior             2          3   
32434211   3093936   205420    prior             1          4   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         7                    15.0         196   
2                        12                    21.0         196   
3                         7                    29.0         196   
4                        15                    28.0         196   
...                     ...                     ...         ...   
32434207                 15                     1.0       43553   
32434208                 11                     3.0       43553   
32434209                  7                     NaN       42338   
32434210                  5                    15.0       42338   
32434211                 14                     NaN       28818   

          add_to_cart_order  reordered _merge  Unnamed: 0  \
0                         1          0   both         195   
1                         1          1   both         195   
2                         1          1   both         195   
3                         1          1   both         195   
4                         1          1   both         195   
...                     ...        ...    ...         ...   
32434207                  2          1   both       43557   
32434208                  2          1   both       43557   
32434209                  4          0   both       42342   
32434210                 16          1   both       42342   
32434211                  8          0   both       28820   

                           product_name  aisle_id  department_id  prices  \
0                                  Soda        77              7     9.0   
1                                  Soda        77              7     9.0   
2                                  Soda        77              7     9.0   
3                                  Soda        77              7     9.0   
4                                  Soda        77              7     9.0   
...                                 ...       ...            ...     ...   
32434207            Orange Energy Shots        64              7     3.7   
32434208            Orange Energy Shots        64              7     3.7   
32434209          Zucchini Chips, Pesto        50             19     6.9   
32434210          Zucchini Chips, Pesto        50             19     6.9   
32434211  Hot Oatmeal Multigrain Raisin       130             14    10.3   

          max_order      loyalty_flag  
0                10      New customer  
1                10      New customer  
2                10      New customer  
3                10      New customer  
4                10      New customer  
...             ...               ...  
32434207         31  Regular customer  
32434208         31  Regular customer  
32434209          3      New customer  
32434210          3      New customer  
32434211         16  Regular customer  

[32434212 rows x 18 columns]

### 3. Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

In [17]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})


order_number
                      mean
department_id             
1                15.457687
2                17.277920
3                17.179756
4                17.811403
5                15.213779
6                16.439806
7                17.225773
8                15.340520
9                15.895474
10               20.197148
11               16.170828
12               15.887622
13               16.583304
14               16.757377
15               16.165037
16               17.663250
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

In [18]:
df.groupby('department_id').agg({'order_number': ['mean']})


order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

Answer: The results of all values in a subset are greater than that of the entire dataframe except the 'department_id' 17

### 4. Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [19]:
# Transform the ‘order_number’ column by applying the max() function from the NumPy library

ords_prods_merge['mean_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [20]:
ords_prods_merge

order_id  user_id eval_set  order_number  order_dow  \
0          2539329        1    prior             1          2   
1          2398795        1    prior             2          3   
2           473747        1    prior             3          3   
3          2254736        1    prior             4          4   
4           431534        1    prior             5          4   
...            ...      ...      ...           ...        ...   
32434207   1320836   202557    prior            17          2   
32434208     31526   202557    prior            18          5   
32434209    758936   203436    prior             1          2   
32434210   2745165   203436    prior             2          3   
32434211   3093936   205420    prior             1          4   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         7                    15.0         196   
2                        12                    21.0         196   
3                         7                    29.0         196   
4                        15                    28.0         196   
...                     ...                     ...         ...   
32434207                 15                     1.0       43553   
32434208                 11                     3.0       43553   
32434209                  7                     NaN       42338   
32434210                  5                    15.0       42338   
32434211                 14                     NaN       28818   

          add_to_cart_order  reordered _merge  Unnamed: 0  \
0                         1          0   both         195   
1                         1          1   both         195   
2                         1          1   both         195   
3                         1          1   both         195   
4                         1          1   both         195   
...                     ...        ...    ...         ...   
32434207                  2          1   both       43557   
32434208                  2          1   both       43557   
32434209                  4          0   both       42342   
32434210                 16          1   both       42342   
32434211                  8          0   both       28820   

                           product_name  aisle_id  department_id  prices  \
0                                  Soda        77              7     9.0   
1                                  Soda        77              7     9.0   
2                                  Soda        77              7     9.0   
3                                  Soda        77              7     9.0   
4                                  Soda        77              7     9.0   
...                                 ...       ...            ...     ...   
32434207            Orange Energy Shots        64              7     3.7   
32434208            Orange Energy Shots        64              7     3.7   
32434209          Zucchini Chips, Pesto        50             19     6.9   
32434210          Zucchini Chips, Pesto        50             19     6.9   
32434211  Hot Oatmeal Multigrain Raisin       130             14    10.3   

          max_order      loyalty_flag  mean_price  
0                10      New customer    6.367797  
1                10      New customer    6.367797  
2                10      New customer    6.367797  
3                10      New customer    6.367797  
4                10      New customer    6.367797  
...             ...               ...         ...  
32434207         31  Regular customer    6.905655  
32434208         31  Regular customer    6.905655  
32434209          3      New customer    7.631579  
32434210          3      New customer    7.631579  
32434211         16  Regular customer    7.684746  

[32434212 rows x 19 columns]

In [21]:
# Create a flag that assigns a “loyalty” label to a user ID based on its corresponding max order value

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)


loyalty_flag
Regular customer    15891077
Loyal customer      10293737
New customer         6249398
Name: count, dtype: int64

### 5. Find whether there’s a difference between the spending habits of the three types of customers; loyal, regular, and new. 
- Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category; loyal, regular, and new customer.
- Determine whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [22]:
# Checking basic statistics on the 'loyalty_flag' column with the values of 'mean', 'min', and 'max' in 'prices'

ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})


prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.388747  1.0  99999.0
New customer      13.294370  1.0  99999.0
Regular customer  12.496203  1.0  99999.0

Answer: The average purchase price of the new (7.805644) customers are highest followed by regular (7.802952) and new loyal (7.778386) customers.

#### Finding and Adjusting abnormal values to get more accurate answer

In [23]:
# Checking descriptive statistics on the 'prices' column

ords_prods_merge['prices'].describe()

count    3.243421e+07
mean     1.198114e+01
std      4.954311e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      9.999900e+04
Name: prices, dtype: float64

##### - Since the maximum purchase price (99999.0) seems abnormal, further analysis will be required.

In [24]:
# Checking for abnormal values in the 'prices' column

ords_prods_merge['prices'].sort_values(ascending = False)


29192983    99999.0
29192963    99999.0
29192971    99999.0
29192970    99999.0
29192969    99999.0
             ...   
23563370        1.0
23563371        1.0
23563372        1.0
23563373        1.0
7882813         1.0
Name: prices, Length: 32434212, dtype: float64

##### - Found that there are a number of abnormal values in the 'price' column which needs to be arranged.

In [25]:
# Dropping duplicate values in the 'prices' column

ords_prods_merge['prices'].drop_duplicates().sort_values(ascending = False)


29192575    99999.0
10030345    14900.0
18529046       25.0
25785942       24.9
28452491       24.8
             ...   
3384024         1.4
632766          1.3
5488887         1.2
5147325         1.1
60725           1.0
Name: prices, Length: 242, dtype: float64

##### - After using 'drop_duplicates,' found another abnormal value (14900.0) in the 'price' column which needs to be arranged.

In [26]:
# Checking the value in the 'prices' column with 99999 or 14900

max_price = ords_prods_merge.loc[ords_prods_merge['prices'].isin([99999, 14900])]
max_price


order_id  user_id eval_set  order_number  order_dow  \
10030345    912404       17    prior            12          2   
10030346    603376       17    prior            22          6   
10030347   3264360      135    prior             2          2   
10030348    892534      135    prior             3          0   
10030349    229704      342    prior             8          1   
...            ...      ...      ...           ...        ...   
29193268   2249946   204099    prior            29          0   
29193269   2363282   204099    prior            31          0   
29193270   3181945   204395    prior            13          3   
29193271   2486215   205227    prior             7          3   
29193272   2232716   205227    prior            10          0   

          order_hour_of_day  days_since_prior_order  product_id  \
10030345                 14                     5.0       21553   
10030346                 16                     4.0       21553   
10030347                 21                    13.0       21553   
10030348                  8                    12.0       21553   
10030349                 19                    30.0       21553   
...                     ...                     ...         ...   
29193268                  8                     4.0       33664   
29193269                  9                     2.0       33664   
29193270                 15                     8.0       33664   
29193271                 20                     4.0       33664   
29193272                 11                     4.0       33664   

          add_to_cart_order  reordered _merge  Unnamed: 0  \
10030345                  5          0   both       21554   
10030346                  3          1   both       21554   
10030347                  6          0   both       21554   
10030348                  3          1   both       21554   
10030349                  9          0   both       21554   
...                     ...        ...    ...         ...   
29193268                  1          0   both       33666   
29193269                  1          1   both       33666   
29193270                 25          0   both       33666   
29193271                  8          0   both       33666   
29193272                 23          1   both       33666   

                              product_name  aisle_id  department_id   prices  \
10030345  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10030346  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10030347  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10030348  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10030349  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
...                                    ...       ...            ...      ...   
29193268             2 % Reduced Fat  Milk        84             16  99999.0   
29193269             2 % Reduced Fat  Milk        84             16  99999.0   
29193270             2 % Reduced Fat  Milk        84             16  99999.0   
29193271             2 % Reduced Fat  Milk        84             16  99999.0   
29193272             2 % Reduced Fat  Milk        84             16  99999.0   

          max_order      loyalty_flag   mean_price  
10030345         40  Regular customer   108.648299  
10030346         40  Regular customer   108.648299  
10030347          4      New customer  1154.792308  
10030348          4      New customer  1154.792308  
10030349         16  Regular customer   114.426619  
...             ...               ...          ...  
29193268         39  Regular customer  1106.743956  
29193269         39  Regular customer  1106.743956  
29193270         15  Regular customer   451.153540  
29193271         12  Regular customer  1178.381871  
29193272         12  Regular customer  1178.381871  

[5127 rows x 19 columns]

In [27]:
# Checking descriptive statistics of the 'max_price'

max_price.describe()


order_id        user_id  order_number    order_dow  \
count  5.127000e+03    5127.000000   5127.000000  5127.000000   
mean   1.708876e+06  106117.185684     15.302126     2.750536   
std    9.838956e+05   58594.982659     16.691849     2.010892   
min    4.300000e+01      17.000000      1.000000     0.000000   
25%    8.642320e+05   58136.000000      4.000000     1.000000   
50%    1.693799e+06  106021.000000      9.000000     3.000000   
75%    2.567244e+06  156389.000000     20.000000     5.000000   
max    3.421047e+06  206049.000000     99.000000     6.000000   

       order_hour_of_day  days_since_prior_order    product_id  \
count        5127.000000             4776.000000   5127.000000   
mean           13.473376               11.899079  23201.815682   
std             4.137326                9.068019   4153.742747   
min             0.000000                0.000000  21553.000000   
25%            10.000000                5.000000  21553.000000   
50%            13.000000                8.000000  21553.000000   
75%            16.000000               16.000000  21553.000000   
max            23.000000               30.000000  33664.000000   

       add_to_cart_order    reordered    Unnamed: 0     aisle_id  \
count        5127.000000  5127.000000   5127.000000  5127.000000   
mean            7.986347     0.613614  23202.951824   104.732592   
std             6.868889     0.486968   4154.085720     8.231346   
min             1.000000     0.000000  21554.000000    84.000000   
25%             3.000000     0.000000  21554.000000   108.000000   
50%             6.000000     1.000000  21554.000000   108.000000   
75%            11.000000     1.000000  21554.000000   108.000000   
max            72.000000     1.000000  33666.000000   108.000000   

       department_id        prices    max_order    mean_price  
count         5127.0   5127.000000  5127.000000   5127.000000  
mean            16.0  26485.547494    29.791691    710.125289  
std              0.0  29186.636448    24.929207   1397.618625  
min             16.0  14900.000000     3.000000     15.508368  
25%             16.0  14900.000000    12.000000    150.997143  
50%             16.0  14900.000000    21.000000    306.956855  
75%             16.0  14900.000000    41.000000    608.500000  
max             16.0  99999.000000    99.000000  25005.425000

##### - Found that the prices of product_id 33664 is 99999, and the prices of product_id 21553 is 14900

In [28]:
# Checking the product_name of product_id 21553 and 33664

ords_prods_merge.loc[ords_prods_merge['product_id'].isin([21553, 33664])]


order_id  user_id eval_set  order_number  order_dow  \
10030345    912404       17    prior            12          2   
10030346    603376       17    prior            22          6   
10030347   3264360      135    prior             2          2   
10030348    892534      135    prior             3          0   
10030349    229704      342    prior             8          1   
...            ...      ...      ...           ...        ...   
29193268   2249946   204099    prior            29          0   
29193269   2363282   204099    prior            31          0   
29193270   3181945   204395    prior            13          3   
29193271   2486215   205227    prior             7          3   
29193272   2232716   205227    prior            10          0   

          order_hour_of_day  days_since_prior_order  product_id  \
10030345                 14                     5.0       21553   
10030346                 16                     4.0       21553   
10030347                 21                    13.0       21553   
10030348                  8                    12.0       21553   
10030349                 19                    30.0       21553   
...                     ...                     ...         ...   
29193268                  8                     4.0       33664   
29193269                  9                     2.0       33664   
29193270                 15                     8.0       33664   
29193271                 20                     4.0       33664   
29193272                 11                     4.0       33664   

          add_to_cart_order  reordered _merge  Unnamed: 0  \
10030345                  5          0   both       21554   
10030346                  3          1   both       21554   
10030347                  6          0   both       21554   
10030348                  3          1   both       21554   
10030349                  9          0   both       21554   
...                     ...        ...    ...         ...   
29193268                  1          0   both       33666   
29193269                  1          1   both       33666   
29193270                 25          0   both       33666   
29193271                  8          0   both       33666   
29193272                 23          1   both       33666   

                              product_name  aisle_id  department_id   prices  \
10030345  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10030346  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10030347  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10030348  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10030349  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
...                                    ...       ...            ...      ...   
29193268             2 % Reduced Fat  Milk        84             16  99999.0   
29193269             2 % Reduced Fat  Milk        84             16  99999.0   
29193270             2 % Reduced Fat  Milk        84             16  99999.0   
29193271             2 % Reduced Fat  Milk        84             16  99999.0   
29193272             2 % Reduced Fat  Milk        84             16  99999.0   

          max_order      loyalty_flag   mean_price  
10030345         40  Regular customer   108.648299  
10030346         40  Regular customer   108.648299  
10030347          4      New customer  1154.792308  
10030348          4      New customer  1154.792308  
10030349         16  Regular customer   114.426619  
...             ...               ...          ...  
29193268         39  Regular customer  1106.743956  
29193269         39  Regular customer  1106.743956  
29193270         15  Regular customer   451.153540  
29193271         12  Regular customer  1178.381871  
29193272         12  Regular customer  1178.381871  

[5127 rows x 19 columns]

##### - Found that the product_name of product_id 21553 is 'Lowfat 2% Milkfat Cottage Cheese,' and the product_name of product_id 33664 is '2 % Reduced Fat Milk'

In [29]:
# Filtering the products with the price of 99999 or 14900 in department_id 16, 'dairy'

dairy = ords_prods_merge.loc[ords_prods_merge['department_id'].isin([16])]
no_abnl = dairy.loc[~dairy['prices'].isin([99999, 14900])]
no_abnl.sample(5)


order_id  user_id eval_set  order_number  order_dow  \
9076253    1680463    53722    prior            34          6   
3870760    2386643   144489    prior            20          5   
29852515   1855002    53862    prior            33          0   
16452109    842378    52430    prior            18          5   
15241493    248812      457    prior             2          0   

          order_hour_of_day  days_since_prior_order  product_id  \
9076253                  12                     3.0       27086   
3870760                  18                     7.0       22035   
29852515                 21                     6.0       24804   
16452109                  9                     7.0       39180   
15241493                 14                     7.0       19348   

          add_to_cart_order  reordered _merge  Unnamed: 0  \
9076253                   4          1   both       27088   
3870760                   4          1   both       22036   
29852515                 17          0   both       24805   
16452109                  1          1   both       39184   
15241493                 14          1   both       19349   

                         product_name  aisle_id  department_id  prices  \
9076253                   Half & Half        53             16    11.4   
3870760   Organic Whole String Cheese        21             16     9.0   
29852515  Original Organic Rice Drink        91             16     3.9   
16452109       Organic Lowfat 1% Milk        84             16     2.1   
15241493                Fat Free Milk        84             16     4.6   

          max_order      loyalty_flag  mean_price  
9076253          83    Loyal customer    7.501677  
3870760          68    Loyal customer    8.324881  
29852515         50    Loyal customer    7.009431  
16452109         24  Regular customer    7.548949  
15241493          4      New customer    8.024194

In [30]:
# Checking the maximum value of the 'prices'

no_abnl['prices'].max()


15.0

##### -  The maximum value of the 'prices' is changed from '99999' to '15' which is considered reasonable.

In [31]:
# To adjust the abnormal price values in 'Lowfat 2% Milkfat Cottage Cheese' (product_id 21553) and '2 % Reduced Fat Milk' (product_id 33664)

# Filtering the products containing '2', 'Milkfat', and 'Cottage Cheese' in their product_name

ctg_chs = no_abnl[no_abnl['product_name'].str.contains("2") & 
                  no_abnl['product_name'].str.contains("Milkfat") & 
                  no_abnl['product_name'].str.contains("Cottage Cheese")]

# Filtering the products containing '2', 'Reduced Fat', and 'Milk' in their product_name

redu_fat_mlk = no_abnl[no_abnl['product_name'].str.contains("2") &
                       no_abnl['product_name'].str.contains("Reduced Fat") &
                       no_abnl['product_name'].str.contains("Milk")]


In [32]:
# Checking ctg_chs

ctg_chs

order_id  user_id eval_set  order_number  order_dow  \
16201300   1030102       52    prior            24          2   
16201301   1588305       52    prior            27          1   
16201302   2362500      123    prior             9          4   
16201303   2649812      345    prior             1          1   
16201304    337387      345    prior             2          6   
...            ...      ...      ...           ...        ...   
32341554   2313427   152072    prior            15          3   
32341555   1826256   152072    prior            17          0   
32341556    631861   152072    prior            19          3   
32341557   2598154   152072    prior            20          4   
32341558    910359   186170    prior             3          6   

          order_hour_of_day  days_since_prior_order  product_id  \
16201300                  9                     8.0       27790   
16201301                  8                    10.0       27790   
16201302                 20                     7.0       27790   
16201303                 13                     NaN       27790   
16201304                 13                    30.0       27790   
...                     ...                     ...         ...   
32341554                  8                    18.0       14065   
32341555                 18                    13.0       14065   
32341556                 17                    15.0       14065   
32341557                 16                    15.0       14065   
32341558                 14                    18.0       14065   

          add_to_cart_order  reordered _merge  Unnamed: 0  \
16201300                  2          0   both       27792   
16201301                  3          1   both       27792   
16201302                  3          0   both       27792   
16201303                  4          0   both       27792   
16201304                  2          1   both       27792   
...                     ...        ...    ...         ...   
32341554                 17          1   both       14065   
32341555                  4          1   both       14065   
32341556                  4          1   both       14065   
32341557                  3          1   both       14065   
32341558                  4          0   both       14065   

                                               product_name  aisle_id  \
16201300        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
16201301        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
16201302        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
16201303        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
16201304        Small Curd Lowfat 2% Milkfat Cottage Cheese       108   
...                                                     ...       ...   
32341554  Cottage Cheese 2% Milkfat Lowfat,Blueberry Top...        21   
32341555  Cottage Cheese 2% Milkfat Lowfat,Blueberry Top...        21   
32341556  Cottage Cheese 2% Milkfat Lowfat,Blueberry Top...        21   
32341557  Cottage Cheese 2% Milkfat Lowfat,Blueberry Top...        21   
32341558  Cottage Cheese 2% Milkfat Lowfat,Blueberry Top...        21   

          department_id  prices  max_order      loyalty_flag  mean_price  
16201300             16     7.7         27  Regular customer    6.408876  
16201301             16     7.7         27  Regular customer    6.408876  
16201302             16     7.7         10      New customer    5.226984  
16201303             16     7.7          3      New customer    6.276923  
16201304             16     7.7          3      New customer    6.276923  
...                 ...     ...        ...               ...         ...  
32341554             16    11.8         20  Regular customer    7.931707  
32341555             16    11.8         20  Regular customer    7.931707  
32341556             16    11.8         20  Regular customer    7.931707  
32341557             16    11.8         20  Regular customer    

In [33]:
# Checking descriptive statistics 

ctg_chs.describe()

order_id        user_id  order_number    order_dow  \
count  3.487000e+03    3487.000000   3487.000000  3487.000000   
mean   1.684106e+06  103778.688271     15.484084     2.849728   
std    9.785618e+05   58583.418992     16.808776     2.031540   
min    2.301000e+03      52.000000      1.000000     0.000000   
25%    8.438150e+05   55839.000000      4.000000     1.000000   
50%    1.673171e+06  107636.000000     10.000000     3.000000   
75%    2.535750e+06  151541.500000     20.000000     5.000000   
max    3.419042e+06  206072.000000     98.000000     6.000000   

       order_hour_of_day  days_since_prior_order    product_id  \
count        3487.000000             3218.000000   3487.000000   
mean           13.257241               12.898073  22348.791225   
std             4.297401                9.231128   8826.332190   
min             0.000000                0.000000    195.000000   
25%            10.000000                6.000000  11203.000000   
50%            13.000000               10.000000  27790.000000   
75%            16.000000               19.000000  27790.000000   
max            23.000000               30.000000  31030.000000   

       add_to_cart_order    reordered    Unnamed: 0     aisle_id  \
count        3487.000000  3487.000000   3487.000000  3487.000000   
mean            5.840551     0.697448  22350.199312   107.650703   
std             5.955836     0.459429   8827.244775     5.502322   
min             1.000000     0.000000    194.000000    21.000000   
25%             2.000000     0.000000  11203.000000   108.000000   
50%             4.000000     1.000000  27792.000000   108.000000   
75%             8.000000     1.000000  27792.000000   108.000000   
max            57.000000     1.000000  31032.000000   108.000000   

       department_id       prices    max_order   mean_price  
count         3487.0  3487.000000  3487.000000  3487.000000  
mean            16.0     7.228219    30.531116    20.097395  
std              0.0     1.757900    25.413902    60.377198  
min             16.0     2.400000     3.000000     3.971429  
25%             16.0     7.700000    11.000000     7.124627  
50%             16.0     7.700000    22.000000     7.688550  
75%             16.0     7.700000    43.000000     8.353303  
max             16.0    11.800000    99.000000   883.061765

##### -  The mean, min, and max values of the 'prices' in the 'ctg_chs' are 7.23, 2.4, and 11.8 respectively which are considered reasonable.

In [34]:
# Checking redu_fat_mlk

redu_fat_mlk

order_id  user_id eval_set  order_number  order_dow  \
8310905    2840752       11    prior             3          5   
8310906    1468214       11    prior             7          5   
8310907    1854701      152    prior             4          4   
8310908    2200108      152    prior             5          1   
8310909    1712459      152    prior             6          5   
...            ...      ...      ...           ...        ...   
32430282   1387229   146914    prior             1          1   
32430283   1555717   146914    prior             2          2   
32430284    783975   146914    prior             4          0   
32430285   2472439   146914    prior             9          2   
32430286   1075865   195045    prior             9          0   

          order_hour_of_day  days_since_prior_order  product_id  \
8310905                  10                    14.0       33572   
8310906                   9                    30.0       33572   
8310907                   9                     9.0       33572   
8310908                  15                    11.0       33572   
8310909                  10                     4.0       33572   
...                     ...                     ...         ...   
32430282                 15                     NaN        4864   
32430283                 14                     8.0        4864   
32430284                 19                     2.0        4864   
32430285                 16                     3.0        4864   
32430286                 17                    20.0        4864   

          add_to_cart_order  reordered _merge  Unnamed: 0  \
8310905                   9          0   both       33574   
8310906                   3          1   both       33574   
8310907                   3          0   both       33574   
8310908                   7          1   both       33574   
8310909                   9          1   both       33574   
...                     ...        ...    ...         ...   
32430282                 16          0   both        4864   
32430283                 12          1   both        4864   
32430284                  6          1   both        4864   
32430285                  1          1   both        4864   
32430286                  5          0   both        4864   

                              product_name  aisle_id  department_id  prices  \
8310905   Lactose Free 2% Reduced Fat Milk        91             16    10.5   
8310906   Lactose Free 2% Reduced Fat Milk        91             16    10.5   
8310907   Lactose Free 2% Reduced Fat Milk        91             16    10.5   
8310908   Lactose Free 2% Reduced Fat Milk        91             16    10.5   
8310909   Lactose Free 2% Reduced Fat Milk        91             16    10.5   
...                                    ...       ...            ...     ...   
32430282   Milk, 2% Reduced Fat, Chocolate        84             16     3.5   
32430283   Milk, 2% Reduced Fat, Chocolate        84             16     3.5   
32430284   Milk, 2% Reduced Fat, Chocolate        84             16     3.5   
32430285   Milk, 2% Reduced Fat, Chocolate        84             16     3.5   
32430286   Milk, 2% Reduced Fat, Chocolate        84             16     3.5   

          max_order      loyalty_flag  mean_price  
8310905           7      New customer    8.328723  
8310906           7      New customer    8.328723  
8310907          65    Loyal customer    7.704454  
8310908          65    Loyal customer    7.704454  
8310909          65    Loyal customer    7.704454  
...             ...               ...         ...  
32430282         26  Regular customer    5.843478  
32430283         26  Regular customer    5.843478  
32430284         26  Regular customer    5.843478  
32430285         26  Regular customer    5.843478  
32430286         12  Regular customer    8.570423  

[202238 rows x 19 columns]

In [35]:
# Checking descriptive statistics 

redu_fat_mlk.describe()

order_id        user_id   order_number      order_dow  \
count  2.022380e+05  202238.000000  202238.000000  202238.000000   
mean   1.711414e+06  103040.384151      18.528684       2.748702   
std    9.876774e+05   59620.791055      18.288088       2.022770   
min    5.000000e+00      11.000000       1.000000       0.000000   
25%    8.586890e+05   51496.000000       5.000000       1.000000   
50%    1.706996e+06  102925.000000      12.000000       3.000000   
75%    2.566178e+06  154743.000000      26.000000       5.000000   
max    3.421036e+06  206209.000000      99.000000       6.000000   

       order_hour_of_day  days_since_prior_order     product_id  \
count      202238.000000           190556.000000  202238.000000   
mean           13.181835               10.538445   18768.761009   
std             4.193794                8.380235   13345.840890   
min             0.000000                0.000000    1940.000000   
25%            10.000000                5.000000    5785.000000   
50%            13.000000                7.000000   19820.000000   
75%            16.000000               14.000000   32478.000000   
max            23.000000               30.000000   49319.000000   

       add_to_cart_order      reordered     Unnamed: 0       aisle_id  \
count      202238.000000  202238.000000  202238.000000  202238.000000   
mean            5.423051       0.782692   18769.890520      83.716626   
std             5.397505       0.412415   13347.113649       6.511429   
min             1.000000       0.000000    1940.000000      21.000000   
25%             2.000000       1.000000    5785.000000      84.000000   
50%             4.000000       1.000000   19821.000000      84.000000   
75%             7.000000       1.000000   32480.000000      84.000000   
max           137.000000       1.000000   49323.000000      91.000000   

       department_id         prices      max_order     mean_price  
count       202238.0  202238.000000  202238.000000  202238.000000  
mean            16.0       9.086900      35.834794      14.774299  
std              0.0       3.064869      26.069087      94.590225  
min             16.0       1.200000       3.000000       2.400000  
25%             16.0       7.900000      14.000000       7.401666  
50%             16.0       9.200000      30.000000       7.861137  
75%             16.0      11.400000      51.000000       8.326024  
max             16.0      14.700000      99.000000   11772.100000

##### -  The mean, min, and max values of the 'prices' in the 'redu_fat_mlk' are 9.09, 1.2, and 14.7 respectively which are considered reasonable.

In [36]:
# Adjest the abnormal values in the 'prices' column

ords_prods_merge['prices'] = ords_prods_merge['prices'].replace({99999: 9.09, 14900: ctg_chs['prices'].mean()})


In [37]:
# Checking descriptive statistics on the 'prices' column

ords_prods_merge['prices'].describe()


count    3.243421e+07
mean     7.795655e+00
std      4.242823e+00
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      2.500000e+01
Name: prices, dtype: float64

In [38]:
# Checking basic statistics on the 'loyalty_flag' column with the values of 'mean', 'min', and 'max' in 'prices'

ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})


prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.778378  1.0  25.0
New customer      7.805615  1.0  25.0
Regular customer  7.802929  1.0  25.0

Answer: The average purchase price of the new (7.805615) customers are highest followed by regular (7.802929) and new loyal (7.778378) customers. Compared to the results of the statistics before abnormal data correction, there is no significant difference, and the ranking is the same.

### 6. Create a spending flag for each user based on the average price across all their orders using the following criteria:
- If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
- If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [40]:
# Rounding 'mean_price' to two decimal points.

ords_prods_merge['mean_price'] = ords_prods_merge['mean_price'].round(2)
ords_prods_merge

order_id  user_id eval_set  order_number  order_dow  \
0          2539329        1    prior             1          2   
1          2398795        1    prior             2          3   
2           473747        1    prior             3          3   
3          2254736        1    prior             4          4   
4           431534        1    prior             5          4   
...            ...      ...      ...           ...        ...   
32434207   1320836   202557    prior            17          2   
32434208     31526   202557    prior            18          5   
32434209    758936   203436    prior             1          2   
32434210   2745165   203436    prior             2          3   
32434211   3093936   205420    prior             1          4   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         7                    15.0         196   
2                        12                    21.0         196   
3                         7                    29.0         196   
4                        15                    28.0         196   
...                     ...                     ...         ...   
32434207                 15                     1.0       43553   
32434208                 11                     3.0       43553   
32434209                  7                     NaN       42338   
32434210                  5                    15.0       42338   
32434211                 14                     NaN       28818   

          add_to_cart_order  reordered _merge  Unnamed: 0  \
0                         1          0   both         195   
1                         1          1   both         195   
2                         1          1   both         195   
3                         1          1   both         195   
4                         1          1   both         195   
...                     ...        ...    ...         ...   
32434207                  2          1   both       43557   
32434208                  2          1   both       43557   
32434209                  4          0   both       42342   
32434210                 16          1   both       42342   
32434211                  8          0   both       28820   

                           product_name  aisle_id  department_id  prices  \
0                                  Soda        77              7     9.0   
1                                  Soda        77              7     9.0   
2                                  Soda        77              7     9.0   
3                                  Soda        77              7     9.0   
4                                  Soda        77              7     9.0   
...                                 ...       ...            ...     ...   
32434207            Orange Energy Shots        64              7     3.7   
32434208            Orange Energy Shots        64              7     3.7   
32434209          Zucchini Chips, Pesto        50             19     6.9   
32434210          Zucchini Chips, Pesto        50             19     6.9   
32434211  Hot Oatmeal Multigrain Raisin       130             14    10.3   

          max_order      loyalty_flag  mean_price  
0                10      New customer        6.37  
1                10      New customer        6.37  
2                10      New customer        6.37  
3                10      New customer        6.37  
4                10      New customer        6.37  
...             ...               ...         ...  
32434207         31  Regular customer        6.91  
32434208         31  Regular customer        6.91  
32434209          3      New customer        7.63  
32434210          3      New customer        7.63  
32434211         16  Regular customer        7.68  

[32434212 rows x 19 columns]

In [48]:
# Creating a 'spending_flag' column and establish the conditions of the flag

ords_prods_merge.loc[ords_prods_merge['mean_price'] < 10, 'spending_flag'] = 'Low spender'
ords_prods_merge.loc[ords_prods_merge['mean_price'] >= 10, 'spending_flag'] = 'High spender'
ords_prods_merge


order_id  user_id eval_set  order_number  order_dow  \
0          2539329        1    prior             1          2   
1          2398795        1    prior             2          3   
2           473747        1    prior             3          3   
3          2254736        1    prior             4          4   
4           431534        1    prior             5          4   
...            ...      ...      ...           ...        ...   
32434207   1320836   202557    prior            17          2   
32434208     31526   202557    prior            18          5   
32434209    758936   203436    prior             1          2   
32434210   2745165   203436    prior             2          3   
32434211   3093936   205420    prior             1          4   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         7                    15.0         196   
2                        12                    21.0         196   
3                         7                    29.0         196   
4                        15                    28.0         196   
...                     ...                     ...         ...   
32434207                 15                     1.0       43553   
32434208                 11                     3.0       43553   
32434209                  7                     NaN       42338   
32434210                  5                    15.0       42338   
32434211                 14                     NaN       28818   

          add_to_cart_order  reordered _merge  Unnamed: 0  \
0                         1          0   both         195   
1                         1          1   both         195   
2                         1          1   both         195   
3                         1          1   both         195   
4                         1          1   both         195   
...                     ...        ...    ...         ...   
32434207                  2          1   both       43557   
32434208                  2          1   both       43557   
32434209                  4          0   both       42342   
32434210                 16          1   both       42342   
32434211                  8          0   both       28820   

                           product_name  aisle_id  department_id  prices  \
0                                  Soda        77              7     9.0   
1                                  Soda        77              7     9.0   
2                                  Soda        77              7     9.0   
3                                  Soda        77              7     9.0   
4                                  Soda        77              7     9.0   
...                                 ...       ...            ...     ...   
32434207            Orange Energy Shots        64              7     3.7   
32434208            Orange Energy Shots        64              7     3.7   
32434209          Zucchini Chips, Pesto        50             19     6.9   
32434210          Zucchini Chips, Pesto        50             19     6.9   
32434211  Hot Oatmeal Multigrain Raisin       130             14    10.3   

          max_order      loyalty_flag  mean_price spending_flag  
0                10      New customer        6.37   Low spender  
1                10      New customer        6.37   Low spender  
2                10      New customer        6.37   Low spender  
3                10      New customer        6.37   Low spender  
4                10      New customer        6.37   Low spender  
...             ...               ...         ...           ...  
32434207         31  Regular customer        6.91   Low spender  
32434208         31  Regular customer        6.91   Low spender  
32434209          3      New customer        7.63   Low spender  
32434210          3      New customer        7.63   Low spender  
32434211         16  Regular customer        7.68   

In [49]:
# Checking the frequency of 'spending_flag'

ords_prods_merge['spending_flag'].value_counts(dropna = False)


spending_flag
Low spender     31798102
High spender      636110
Name: count, dtype: int64

### 7. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
- If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
- If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
- If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [50]:
# Creating a 'spending_flag' column and establish the conditions of the flag

ords_prods_merge.loc[ords_prods_merge['days_since_prior_order'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'
ords_prods_merge.loc[(ords_prods_merge['days_since_prior_order'] > 10) & (ords_prods_merge['days_since_prior_order'] <= 20), 'order_frequency_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['days_since_prior_order'] <= 10, 'order_frequency_flag'] = 'Frequent customer'
ords_prods_merge

order_id  user_id eval_set  order_number  order_dow  \
0          2539329        1    prior             1          2   
1          2398795        1    prior             2          3   
2           473747        1    prior             3          3   
3          2254736        1    prior             4          4   
4           431534        1    prior             5          4   
...            ...      ...      ...           ...        ...   
32434207   1320836   202557    prior            17          2   
32434208     31526   202557    prior            18          5   
32434209    758936   203436    prior             1          2   
32434210   2745165   203436    prior             2          3   
32434211   3093936   205420    prior             1          4   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         7                    15.0         196   
2                        12                    21.0         196   
3                         7                    29.0         196   
4                        15                    28.0         196   
...                     ...                     ...         ...   
32434207                 15                     1.0       43553   
32434208                 11                     3.0       43553   
32434209                  7                     NaN       42338   
32434210                  5                    15.0       42338   
32434211                 14                     NaN       28818   

          add_to_cart_order  reordered  ... Unnamed: 0  \
0                         1          0  ...        195   
1                         1          1  ...        195   
2                         1          1  ...        195   
3                         1          1  ...        195   
4                         1          1  ...        195   
...                     ...        ...  ...        ...   
32434207                  2          1  ...      43557   
32434208                  2          1  ...      43557   
32434209                  4          0  ...      42342   
32434210                 16          1  ...      42342   
32434211                  8          0  ...      28820   

                           product_name aisle_id  department_id  prices  \
0                                  Soda       77              7     9.0   
1                                  Soda       77              7     9.0   
2                                  Soda       77              7     9.0   
3                                  Soda       77              7     9.0   
4                                  Soda       77              7     9.0   
...                                 ...      ...            ...     ...   
32434207            Orange Energy Shots       64              7     3.7   
32434208            Orange Energy Shots       64              7     3.7   
32434209          Zucchini Chips, Pesto       50             19     6.9   
32434210          Zucchini Chips, Pesto       50             19     6.9   
32434211  Hot Oatmeal Multigrain Raisin      130             14    10.3   

          max_order      loyalty_flag mean_price  spending_flag  \
0                10      New customer       6.37    Low spender   
1                10      New customer       6.37    Low spender   
2                10      New customer       6.37    Low spender   
3                10      New customer       6.37    Low spender   
4                10      New customer       6.37    Low spender   
...             ...               ...        ...            ...   
32434207         31  Regular customer       6.91    Low spender   
32434208         31  Regular customer       6.91    Low spender   
32434209          3      New customer       7.63    Low spender   
32434210          3      New customer       7.63    Low spender   
32434211         16  Regular customer       7.68    Low spender   

           order_freq

In [52]:
# Checking the frequency of 'order_frequency_flag'

ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)


order_frequency_flag
Frequent customer        19059612
Regular customer          6061360
Non-frequent customer     5235156
NaN                       2078084
Name: count, dtype: int64

### 8. Ensure your notebook is clean and structured and that your code is well commented.

Okay!

### 9. Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.¶

In [54]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_updated_090723.pkl'))
